lib/resources/packages.rb

```diff
@@ -0,0 +1,86 @@
+# encoding: utf-8
+# copyright: 2017, Chef Software, Inc. <legal@chef.io>
+# author: Joshua Timberman
+# author: Alex Pop
+# license: All rights reserved
+
+require 'utils/filter'
+
+module Inspec::Resources
+  class Packages < Inspec.resource(1)
+    name 'packages'
+    desc 'Use the packages InSpec audit resource to test properties for multiple packages installed on the system'
+    example "
+      describe packages('vim*').entries.length do
+        it { should be > 1 }
+      end
+      describe packages(/.+/).where { status != 'installed' } do
+        its('statuses') { should be_empty }
+      end
+    "
+
+    def initialize(pattern)
+      @pattern = pattern_regexp(pattern)
+      all_pkgs = package_list
+      @list = all_pkgs.find_all do |hm|
+        hm[:name] =~ pattern_regexp(pattern)
+      end
+    end
+
+    def to_s
+      "Packages #{@pattern.class == String ? @pattern : @pattern.inspect}"
+    end
+
+    filter = FilterTable.create
+    filter.add_accessor(:where)
+          .add_accessor(:entries)
+          .add(:statuses,  field: 'status', style: :simple)
+          .add(:names,     field: 'name')
+          .add(:versions,  field: 'version')
+          .connect(self, :filtered_packages)
+
+    private
+
+    # dpkg takes a shell glob, like 'vim*', but we're filtering using
+    # a regular expression. let's make sure that the pattern gets
+    # converted to a regular expression
+    def pattern_regexp(p)
+      shmap = { '*'=>'.*', '?' => '.', '[' => '[', ']' => ']' }
+      if p.class == String
+        pat = p.gsub(/(.)/) { |m| shmap[m] || m }
+        Regexp.new('^' + pat + '(\s|$)')
+      else
+        p
+      end
+    end
+
+    def filtered_packages
+      @list
+    end
+
+    def package_list
+      os = inspec.os
+
+      if os.debian?
+        command = "dpkg-query -W -f='${db:Status-Abbrev} ${Package} ${Version}\\n'"
+      else
+        fail "packages resource is not yet supported on #{os.name}"
+      end
+      build_package_list(command)
+    end
+
+    Package = Struct.new(:status, :name, :version)
+
+    def build_package_list(command)
+      cmd = inspec.command(command)
+      all = cmd.stdout.split("\n")[1..-1]
+      return [] if all.nil?
+      all.map do |m|
+        a = m.split
+        a[0] = 'installed' if a[0] =~ /^ii/
```

Comment:
Skip the desired action for now. Let's focus on the actual state of the package. We will keep them separate in the future. I.e.:

```
match(/^([uihrp])([ncHUFWti])\s/)
=> res[1] we will ignore for now, it is part of a future field for desired state
=> res[2] is the status, we support installed right now
```

We will establish `installed` as a shared status across all OS. The others are ok to be os-specific for now. imho implement as lowercase with spaces, e.g. `not installed`. Do not export the other status types for now, just stick to installed + open the others in a RFC. Sync with the package resource.